qa/tasks/deepsea_deploy.py

```diff
@@ -249,19 +327,184 @@ def _master_python_version(self, py_version):
+        else:
+            ConfigError(
+                "deepsea_deploy post-Stage 1: unknown profile ->{}<-"
+                .format(profile)
+                )
+
+    def __policy_cfg_build_profile_x(self, profile):
+        self.profile_ymls_to_dump = []
+        no_osd_roles = ("deepsea_deploy: no osd roles configured"
+                        ", but at least one of these is required.")
+        role_dict = {}
+        try:
+            role_dict = self.role_lookup_table['osd']
+        except KeyError:
+            raise ConfigError(no_osd_roles)
+        self.log.debug((
+            "deepsea_deploy: generating policy.cfg lines for osd "
+            "profile ->{}<- based on {}"
+            ).format(profile, role_dict))
+        if len(role_dict.keys()) < 1:
+            raise ConfigError(no_osd_roles)
+        osd_remotes = list(set(role_dict.values()))
+        for osd_remote in osd_remotes:
+            self.log.debug("deepsea_deploy: {} has one or more osd roles"
+                           .format(osd_remote))
+            self.policy_cfg += """# Storage profile - {remote}
+profile-{profile}/cluster/{remote}.sls
+""".format(remote=osd_remote, profile=profile)
+            ypp = ("profile-{}/stack/default/ceph/minions/{}.yml"
+                   .format(profile, osd_remote))
+            self.policy_cfg += ypp + "\n"
+            self.profile_ymls_to_dump.append(
+                "{}/{}".format(proposals_dir, ypp))
+
+    def _policy_cfg_build_x(self, role_type, required=False):
+        no_roles_of_type = ("deepsea_deploy: no {} roles configured"
+                            .format(role_type))
+        but_required = ", but at least one of these is required."
+        role_dict = {}
+        try:
+            role_dict = self.role_lookup_table[role_type]
+        except KeyError:
+            if required:
+                raise ConfigError(no_roles_of_type + but_required)
+            else:
+                self.log.debug(no_roles_of_type)
+                return None
+        self.log.debug(
+            "deepsea_deploy: generating policy.cfg lines for {} based on {}"
+            .format(role_type, role_dict)
+        )
+        if required:
+            if len(role_dict.keys()) < 1:
+                raise ConfigError(no_roles_of_type + but_required)
+        for role_spec, remote_name in role_dict.iteritems():
+            self.log.debug("deepsea_deploy: {} role ->{}<-"
+                           .format(role_type, role_spec))
+            self.log.debug("deepsea_deploy: {} remote name ->{}<-"
+                           .format(role_type, remote_name))
+            self.policy_cfg += """# Role assignment - {role}
+role-{role_type}/cluster/{remote}.sls
+""".format(role=role_spec, role_type=role_type, remote=remote_name)
+
+    def _policy_cfg_cat(self):
+        """
+        Dump the remote policy.cfg file to teuthology log.
+        """
+        self.master_remote.run(args=[
```

Comment:
The problem I had with this was I have not seen any upstream code that passes a string value as the `args` parameter to `run`. Always it is an array - presumably to take advantage of the quoting mechanism.

I think using a string is fine, as long as the quoting mechanism is not needed. It also makes it easier to grep the codebase for these commands.

So, yes, I'll use string instead of list for `args` where feasible.